nmpc_examples/mhe/mhe_constructor.py

```diff
@@ -0,0 +1,269 @@
+        #     else:
+        #         raise KeyError(i)
+
+        if not con_ele.equality:
+            raise RuntimeError(con_ele.name, " is not an equality constraint.")
+        if con_ele.lower.value != con_ele.upper.value:
+            raise RuntimeError(con_ele_name, " has different lower and upper "
+                               "bounds.")
+        if con_ele.upper.value != 0.0:
+            raise RuntimeError(con_ele_name, " is an equality but its bound "
+                               "is not zero.")
+
+        spt = curr_sample_point(i, sample_points)
+
+        return con_ele.body + disturbance_var[index, spt] == \
+            con_ele.upper.value
+    disturbed_con = Constraint(
+        disturbance_set, time, rule=disturbed_con_rule,
+    )
+
+    return disturbance_set, disturbance_var, disturbed_con
+
+
+def activate_disturbed_constraints_based_on_original_constraints(
+    time,
+    sample_points,
+    disturbance_var,
+    mod_constraints,
+    disturbed_con,
+):
+    """
+    This function activate the model constraint and deactivate the original
+    constraint, if the original constarint is active. Also, if all model
+    constraints within a specific sample period are not active, fix the
+    disturbance variable at zero.
+
+    Parameters
+    ----------
+    time: iterable
+        Time set which indexes model constraints
+    sample_points: iterable
+        Set of sample points
+    disturbance_var: Pyomo Var
+        Model disturbances
+    mod_constraints: list
+        List of original model constraints
+    disturbed_con: Pyomo Constraint
+        Model constraints with model disturbances
+
+    """
+    # Deactivate original equalities and activate disturbed equalities
+    for index, con in enumerate(mod_constraints):
+        for tp in time:
+            con_ele = con[tp]
+            if con_ele.active:
+                con_ele.deactivate()
+                disturbed_con[index, tp].activate()
+            else:
+                # If the original equality is not active, deactivate the
+                # disturbed one.
+                disturbed_con[index, tp].deactivate()
+
+    # If all constraints in a specific sample time are not active, fix that
+    # model disturbance at zero.
+    spt_saw_tp = {spt:[] for spt in sample_points}
+    for tp in time:
+        spt = curr_sample_point(tp, sample_points)
+        spt_saw_tp[spt].append(tp)
+
+    for index in range(len(mod_constraints)):
+        for spt in sample_points:
+            con_active_list = [
+                disturbed_con[index, tp].active for tp in spt_saw_tp[spt]
+            ]
+            if not any(con_active_list):
+                disturbance_var[index, spt].fix(0.0)
+
+
+def get_error_disturbance_cost(
+    time,
+    sample_points,
+    components,
+    error_dist_var,
+    weight_data=None,
+):
+    """
+    #TODO: This function is similar to "get_tracking_cost_from_constant_setpoint",
+    #Should I merge this one to that one?
```

Comment:
Cool, I should have thought of setting the setpoint data to zeros.
Just a minor suggestion, shall we remove "Tracking" in line 69 of `nmpc_examples/nmpc/cost_expressions.py` because MHE is using this function as well?

Reply:
I also see you have a toggle between two different objective formulations for constructing measurement error costs. It looks good.